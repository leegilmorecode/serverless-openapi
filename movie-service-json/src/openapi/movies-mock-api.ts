import { movies } from '../data';
import { version } from '../shared/open-api-info/schema-version';

export const getMockApiJson = () => {
  return {
    openapi: '3.0.3',
    'x-amazon-apigateway-request-validators': {
      validation: {
        validateRequestBody: true,
        validateRequestParameters: true,
      },
    },
    info: {
      title: 'Movie Mock API Example',
      version,
      description: 'A Mock API for movies',
    },
    paths: {
      [`/${version}/movies`]: {
        get: {
          'x-amazon-apigateway-request-validator': 'validation',
          summary: 'Get all movies',
          responses: {
            '200': {
              description: 'Successful response',
              content: {
                'application/json': {
                  schema: {
                    type: 'array',
                    items: { $ref: '#/components/schemas/Movie' },
                  },
                },
              },
            },
            '400': {
              description: 'Bad Request',
              content: {
                'application/json': {
                  schema: { $ref: '#/components/schemas/ErrorResponse' },
                },
              },
            },
            '404': {
              description: 'Not Found',
              content: {
                'application/json': {
                  schema: { $ref: '#/components/schemas/ErrorResponse' },
                },
              },
            },
          },
          'x-amazon-apigateway-integration': {
            type: 'mock',
            requestTemplates: {
              'application/json': '{"statusCode": 200}',
            },
            responses: {
              default: {
                statusCode: '200',
                responseTemplates: {
                  'application/json': JSON.stringify(movies),
                },
              },
            },
          },
        },
        post: {
          'x-amazon-apigateway-request-validator': 'validation',
          summary: 'Create a new movie',
          requestBody: {
            required: true,
            content: {
              'application/json': {
                schema: { $ref: '#/components/schemas/NewMovie' },
              },
            },
          },
          responses: {
            '200': {
              description: 'Successful response',
              content: {
                'application/json': {
                  schema: { $ref: '#/components/schemas/Movie' },
                },
              },
            },
            '400': {
              description: 'Bad Request',
              content: {
                'application/json': {
                  schema: { $ref: '#/components/schemas/ErrorResponse' },
                },
              },
            },
            '404': {
              description: 'Not Found',
              content: {
                'application/json': {
                  schema: { $ref: '#/components/schemas/ErrorResponse' },
                },
              },
            },
          },
          'x-amazon-apigateway-integration': {
            type: 'mock',
            requestTemplates: {
              'application/json':
                '#set($context.requestOverride.path.body = $input.body)\n{\n  "statusCode": 200,\n}',
            },
            responses: {
              default: {
                statusCode: '200',
                responseTemplates: {
                  'application/json':
                    '#set($body = $util.parseJson($context.requestOverride.path.body))\n{"id": "c4887ba4-0782-471c-bddc-af50265c96b9",\n "title": "$body.title",\n "rating": "$body.rating",\n "year": "$body.year"\n}',
                },
              },
            },
          },
        },
      },
      [`/${version}/movies/{id}`]: {
        get: {
          'x-amazon-apigateway-request-validator': 'validation',
          summary: 'Get a movie by id',
          parameters: {
            name: 'id',
            in: 'path',
            description: 'ID of the movie',
            required: true,
            schema: { type: 'string' },
          },
          responses: {
            '200': {
              description: 'Successful response',
              content: {
                'application/json': {
                  schema: { $ref: '#/components/schemas/Movie' },
                },
              },
            },
            '400': {
              description: 'Bad Request',
              content: {
                'application/json': {
                  schema: { $ref: '#/components/schemas/ErrorResponse' },
                },
              },
            },
            '404': {
              description: 'Not Found',
              content: {
                'application/json': {
                  schema: { $ref: '#/components/schemas/ErrorResponse' },
                },
              },
            },
          },
          'x-amazon-apigateway-integration': {
            type: 'mock',
            requestTemplates: {
              'application/json': '{"statusCode": 200}',
            },
            responses: {
              default: {
                statusCode: '200',
                responseTemplates: {
                  'application/json': JSON.stringify(movies[0]),
                },
              },
            },
          },
        },
      },
    },
    components: {
      schemas: {
        ErrorResponse: {
          type: 'object',
          required: ['message'],
          properties: { message: { type: 'string' } },
        },
        NewMovie: {
          type: 'object',
          required: ['title', 'year', 'rating'],
          properties: {
            title: {
              type: 'string',
              pattern: '^[a-zA-Z0-9 ]*$',
              minLength: 1,
              maxLength: 100,
              description:
                'The movie title (alphanumeric characters and spaces only)',
            },
            year: {
              type: 'string',
              pattern: '^\\d{4}$',
              description: 'The release year of the movie',
            },
            rating: {
              type: 'string',
              enum: ['U', 'PG', '12', '15', '18'],
              pattern: '^[UPG]|1[258]$',
              description: 'The rating of the movie',
            },
          },
        },
        Movie: {
          type: 'object',
          required: ['id', 'title', 'year', 'rating'],
          properties: {
            id: {
              type: 'string',
              pattern:
                '^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-4[0-9a-fA-F]{3}-[89aAbB][0-9a-fA-F]{3}-[0-9a-fA-F]{12}$',
              description: 'The movie ID (numeric characters only)',
            },
            title: {
              type: 'string',
              pattern: '^[a-zA-Z0-9 ]*$',
              minLength: 1,
              maxLength: 100,
              description:
                'The movie title (alphanumeric characters and spaces only)',
            },
            year: {
              type: 'string',
              pattern: '^\\d{4}$',
              description: 'The release year of the movie',
            },
            rating: {
              type: 'string',
              enum: ['U', 'PG', '12', '15', '18'],
              pattern: '^[UPG]|1[258]$',
              description: 'The rating of the movie',
            },
          },
        },
      },
    },
  };
};
