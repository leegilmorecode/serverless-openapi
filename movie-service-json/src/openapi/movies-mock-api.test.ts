import { getMockApiJson } from './movies-mock-api';

describe('mock-api', () => {
  it('should return the correct mock open api json', () => {
    expect(getMockApiJson()).toMatchSnapshot();
  });
});
