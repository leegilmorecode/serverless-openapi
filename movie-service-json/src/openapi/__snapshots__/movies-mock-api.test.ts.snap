// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mock-api should return the correct mock open api json 1`] = `
{
  "components": {
    "schemas": {
      "ErrorResponse": {
        "properties": {
          "message": {
            "type": "string",
          },
        },
        "required": [
          "message",
        ],
        "type": "object",
      },
      "Movie": {
        "properties": {
          "id": {
            "description": "The movie ID (numeric characters only)",
            "pattern": "^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-4[0-9a-fA-F]{3}-[89aAbB][0-9a-fA-F]{3}-[0-9a-fA-F]{12}$",
            "type": "string",
          },
          "rating": {
            "description": "The rating of the movie",
            "enum": [
              "U",
              "PG",
              "12",
              "15",
              "18",
            ],
            "pattern": "^[UPG]|1[258]$",
            "type": "string",
          },
          "title": {
            "description": "The movie title (alphanumeric characters and spaces only)",
            "maxLength": 100,
            "minLength": 1,
            "pattern": "^[a-zA-Z0-9 ]*$",
            "type": "string",
          },
          "year": {
            "description": "The release year of the movie",
            "pattern": "^\\d{4}$",
            "type": "string",
          },
        },
        "required": [
          "id",
          "title",
          "year",
          "rating",
        ],
        "type": "object",
      },
      "NewMovie": {
        "properties": {
          "rating": {
            "description": "The rating of the movie",
            "enum": [
              "U",
              "PG",
              "12",
              "15",
              "18",
            ],
            "pattern": "^[UPG]|1[258]$",
            "type": "string",
          },
          "title": {
            "description": "The movie title (alphanumeric characters and spaces only)",
            "maxLength": 100,
            "minLength": 1,
            "pattern": "^[a-zA-Z0-9 ]*$",
            "type": "string",
          },
          "year": {
            "description": "The release year of the movie",
            "pattern": "^\\d{4}$",
            "type": "string",
          },
        },
        "required": [
          "title",
          "year",
          "rating",
        ],
        "type": "object",
      },
    },
  },
  "info": {
    "description": "A Mock API for movies",
    "title": "Movie Mock API Example",
    "version": "v1",
  },
  "openapi": "3.0.3",
  "paths": {
    "/v1/movies": {
      "get": {
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "items": {
                    "$ref": "#/components/schemas/Movie",
                  },
                  "type": "array",
                },
              },
            },
            "description": "Successful response",
          },
          "400": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/ErrorResponse",
                },
              },
            },
            "description": "Bad Request",
          },
          "404": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/ErrorResponse",
                },
              },
            },
            "description": "Not Found",
          },
        },
        "summary": "Get all movies",
        "x-amazon-apigateway-integration": {
          "requestTemplates": {
            "application/json": "{"statusCode": 200}",
          },
          "responses": {
            "default": {
              "responseTemplates": {
                "application/json": "[{"id":"180aba94-88e8-4f83-9dcb-0dd437d93ff8","title":"movie one","year":"2019","rating":"18"},{"id":"cbbccaba-3ab9-4af0-a157-0e5cfd654e20","title":"movie two","year":"2020","rating":"PG"},{"id":"31adad13-c876-4231-b8b1-cb45926ba1c8","title":"movie three","year":"2007","rating":"U"}]",
              },
              "statusCode": "200",
            },
          },
          "type": "mock",
        },
        "x-amazon-apigateway-request-validator": "validation",
      },
      "post": {
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "$ref": "#/components/schemas/NewMovie",
              },
            },
          },
          "required": true,
        },
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/Movie",
                },
              },
            },
            "description": "Successful response",
          },
          "400": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/ErrorResponse",
                },
              },
            },
            "description": "Bad Request",
          },
          "404": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/ErrorResponse",
                },
              },
            },
            "description": "Not Found",
          },
        },
        "summary": "Create a new movie",
        "x-amazon-apigateway-integration": {
          "requestTemplates": {
            "application/json": "#set($context.requestOverride.path.body = $input.body)
{
  "statusCode": 200,
}",
          },
          "responses": {
            "default": {
              "responseTemplates": {
                "application/json": "#set($body = $util.parseJson($context.requestOverride.path.body))
{"id": "c4887ba4-0782-471c-bddc-af50265c96b9",
 "title": "$body.title",
 "rating": "$body.rating",
 "year": "$body.year"
}",
              },
              "statusCode": "200",
            },
          },
          "type": "mock",
        },
        "x-amazon-apigateway-request-validator": "validation",
      },
    },
    "/v1/movies/{id}": {
      "get": {
        "parameters": {
          "description": "ID of the movie",
          "in": "path",
          "name": "id",
          "required": true,
          "schema": {
            "type": "string",
          },
        },
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/Movie",
                },
              },
            },
            "description": "Successful response",
          },
          "400": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/ErrorResponse",
                },
              },
            },
            "description": "Bad Request",
          },
          "404": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/ErrorResponse",
                },
              },
            },
            "description": "Not Found",
          },
        },
        "summary": "Get a movie by id",
        "x-amazon-apigateway-integration": {
          "requestTemplates": {
            "application/json": "{"statusCode": 200}",
          },
          "responses": {
            "default": {
              "responseTemplates": {
                "application/json": "{"id":"180aba94-88e8-4f83-9dcb-0dd437d93ff8","title":"movie one","year":"2019","rating":"18"}",
              },
              "statusCode": "200",
            },
          },
          "type": "mock",
        },
        "x-amazon-apigateway-request-validator": "validation",
      },
    },
  },
  "x-amazon-apigateway-request-validators": {
    "validation": {
      "validateRequestBody": true,
      "validateRequestParameters": true,
    },
  },
}
`;
